### PR TITLE
ENH: add wavedecn and waverecn functions

### PR DESCRIPTION
**Pull request by grlee77**
adds n-dimensional variants of wavedec2 and waverec2.
This was split off from the now closed #52

### CODE REVIEW

File: pywt/multilevel.py [hunk 1]
```diff
@@ -220,3 +220,153 @@ def waverec2(coeffs, wavelet, mode='sym'):
...
+    ----------
+    data : ndarray
+        nD input data
+    wavelet : Wavelet object or name string
+        Wavelet to use
+    mode : str, optional
+        Signal extension mode, see MODES (default: 'sym')
+    level : int, optional
+        Decomposition level. If level is None (default) then it will be
+        calculated using the ``dwt_max_level`` function.
```
Comment (rgommers) [inline pywt/multilevel.py]: Would be good to document here that level needs to be >=0

File: pywt/multilevel.py [hunk 2]
```diff
@@ -220,3 +220,153 @@ def waverec2(coeffs, wavelet, mode='sym'):
...
+    mode : str, optional
+        Signal extension mode, see MODES (default: 'sym')
+
+    Returns
+    -------
+    nD array of reconstructed data.
+
+    Examples
+    --------
+    >>> from pywt import multilevel
```
Comment (rgommers) [inline pywt/multilevel.py]: All the functions in this file are also available in the ``pywt`` namespace, so I'm thinking that it would be better to remove ``multilevel`` completely as something that looks like a public submodule (by underscoring the name). In the examples using ``pywt.waverecn`` would be cleaner anyway I think.

**Comment by kwohlfahrt** [comment]
Looks good at a quick glance. This is one way of working around #54. Personally, I think `idwtn` should be strict in what input shapes it accepts, and not silently truncate it's input, but for a multilevel transform this is the correct thing to do. Should we then close #54, or still change the behaviour of `idwtn`?

**Comment by grlee77** [comment]
If you want to add documentation on why the behaviour in #54 is expected and mention this as a workaround, I would be okay with that.  I don't really know which behaviour is preferable for ``idwtn``,

**Comment by grlee77** [comment]
This was rebased using current master.  A major overhaul of the tests was done in the last commit to make them more comprehensive and consistent across all dimensions.

**Comment by grlee77** [comment]
@kwohlfahrt can you check if the updated implementation of `waverecn` here looks reasonable to you.  I added a utility function called `_match_coeff_dims` to get things to work now that `take` is no longer an argument of `idwtn`.

**Comment by kwohlfahrt** [comment]
@grlee77 sorry, I completely missed that I'd been tagged here. The idea looks clean, however in `idwtn` we allow (and silently ignore) the dictionary to contain keys which are not just combinations of `'a'` and `'d'`, and values which are `None`. `_match_coeff_dims` does not handle this at the moment.

There are a few edge cases (e.g. one level being `{}`) that are not handled as well. I'll write tests and fixes for as many as I can catch and open a PR on your repo. Sorry again for taking so long to respond.

**Comment by grlee77** [comment]
I merged @kwolfahrt's code to handle missing or invalid coefficients, but please do not merge just yet.  I still want to add a warning in the case of missing coefficients or unexpected extra keys in the detail coefficient dictionary and an exception if any coefficients have the wrong shape.

**Comment by grlee77** [comment]
@kwohlfahrt 
Please review the most recent commit.  I added an `enable_warnings` argument to `_fix_coeffs` that raises errors for the following two cases:

1.  any detail coefficient arrays are None
2.  any keys in the detail coefficient dictionary have invalid names

`waverecn` itself now raises a warning if any of the detail coefficient dictionaries are empty.

`idwtn` also uses `fix_coeffs`, but I did not enable the warnings there yet.  Do you think we should do so?  If so, perhaps `enable_warnings` should be False in the `waverecn` call to _fix_coeffs so we don't raise the same warnings twice (or the call to `_fix_coeffs` could be removed from `waverecn` since `idwtn` is going to call it again anyways).

**Comment by kwohlfahrt** [comment]
Sorry for the delay, here are my thoughts:

1. I think it's probably better to make this an exception, and say all values must be `array_like`
  - If we're worried about compatibility, a deprecation warning for a release cycle is probably OK. Though iirc we were breaking back-compat so no need to fuss. I'd catch the exception and raise the warning instead, just to get the code for the future written now.
2. Maybe this too? I'm pretty on the fence about it.

I'm not sure it's necessary to warn if any detail coefficient dictionaries are empty. It's an extension of allowing any other number of coefficients to be missing, so it's not really unusual enough to cause a warning.

I would also enable the warnings/exceptions for `idwtn`. If it is used everywhere, I don't think the `enable_warnings` flag should exist, it should just be the behaviour of the function.

**Comment by grlee77** [comment]
@kwohlfahrt 
see how this looks and let me know if there is a cleaner way to do this.  

I removed the warning about empty coefficient dictionaries as suggested.

I now raise a `DeprecationWarning` if any dictionary entries are `None` or have an invalid key name.  This will become a `ValueError` after the deprecation period.

**Comment by kwohlfahrt** [comment]
`_fix_coeffs` should also check if there are (valid) keys that are not the same length. A dictionary containing 'aa' and 'aaa' should be an error.

Given that we 'broke' `idwtn` already in #125, I don't think the deprecation warning is necessary any more, but up to you.

Otherwise looks good!

**Comment by grlee77** [comment]
@kwohlfahrt 
any other comments here?  If not, feel free to merge it

File: pywt/_multilevel.py [hunk 3]
```diff
@@ -397,3 +413,174 @@ def iswt2(coeffs, wavelet):
...
+    a = data
+    for i in range(level):
+        coeffs = dwtn(a, wavelet, mode)
+        a = coeffs.pop('a' * data.ndim)
+        coeffs_list.append(coeffs)
+
+    coeffs_list.append(a)
+    coeffs_list.reverse()
+
+    if len(coeffs_list) < level + 1:
```
Comment (kwohlfahrt) [inline pywt/_multilevel.py]: Looking at the lines above, this test shouldn't ever fail (and it doesn't break any tests if removed).

File: pywt/_multilevel.py [hunk 4]
```diff
@@ -397,3 +413,174 @@ def iswt2(coeffs, wavelet):
...
+        raise ValueError("Expected sequence of coefficient arrays.")
+
+    if len(coeffs) < 1:
+        raise ValueError(
+            "Coefficient list too short (minimum 1 array required).")
+
+    a, ds = coeffs[0], coeffs[1:]
+
+    # Raise error for invalid key combinations
+    ds = list(map(_fix_coeffs, ds))
```
Comment (kwohlfahrt) [inline pywt/_multilevel.py]: Also need need to test that the length of all coeffs are equal to `a.ndim`, otherwise the output has the wrong shape.

Comment (grlee77) [inline pywt/_multilevel.py]: good catch.  I have implemented this now, but after rebasing off of current master there are some new test failures.  I think they are all related to the fact that  `idwt2` now calls `idwtn` so I will have to modify some of those tests that used None for the coefficients, etc.

**Comment by grlee77** [comment]
I ended up not needing to modify any of the existing 2D tests.  I just had to add a line to `idwt2` to drop any coefficients that were None before calling `idwtn`.

**Comment by grlee77** [comment]
@kwohlfahrt : this is now ready to review again

File: pywt/_multidim.py [hunk 5]
```diff
@@ -147,9 +151,43 @@ def dwtn(data, wavelet, mode='symmetric'):
...
+
+    key_lengths = [len(k) for k in coeffs.keys()]
+    if len(np.unique(key_lengths)) > 1:
+        raise ValueError(
+            "All detail coefficient names must have equal length.")
+
+    # make sure the entries of coeffs are all numpy arrays
+    coeffs = dict((k, np.asarray(v)) for k, v in coeffs.items())
+
+    ndim = np.unique(key_lengths)
```
Comment (kwohlfahrt) [inline pywt/_multidim.py]: This returns an array of length one (we've already checked above to make sure there is only one unique value). I'd check in

Comment (grlee77) [inline pywt/_multidim.py]: right.  I went ahead and merged #130 and then will fix this after I rebase again (hopefully for the last time!).

**Comment by grlee77** [comment]
@kwohlfahrt:  still waiting on Travis, but I addressed the issue above and added a couple new tests of the axes arguments to `dwtn`, `idwtn`.

File: pywt/_multilevel.py [hunk 6]
```diff
@@ -399,3 +415,182 @@ def iswt2(coeffs, wavelet):
...
+        [ 1.,  1.,  1.,  1.]],
+
+       [[ 1.,  1.,  1.,  1.],
+        [ 1.,  1.,  1.,  1.],
+        [ 1.,  1.,  1.,  1.],
+        [ 1.,  1.,  1.,  1.]]])
+    """
+
+    if not isinstance(coeffs, (list, tuple)):
+        raise ValueError("Expected sequence of coefficient arrays.")
```
Comment (kwohlfahrt) [inline pywt/_multilevel.py]: I don't think this check is necessary - other sequences should work fine.

**Comment by kwohlfahrt** [comment]
Apart from the `isinstance` check, looks good to me. I don't think we should reject other types of sequence unless there's a particular reason for this.

**Comment by grlee77** [comment]
okay.  should be good to go now
